src/shared/user-record.c

```diff
@@ -21,6 +23,98 @@
 #define DEFAULT_RATELIMIT_BURST 30
 #define DEFAULT_RATELIMIT_INTERVAL_USEC (1*USEC_PER_MINUTE)
 
+#if ENABLE_COMPAT_MUTABLE_UID_BOUNDARIES
+static int parse_alloc_uid(const char *path, const char *name, const char *t, uid_t *ret_uid) {
+        uid_t uid;
+        int r;
+
+        r = parse_uid(t, &uid);
+        if (r < 0)
+                return log_debug_errno(r, "%s: failed to parse %s %s, ignoring: %m", path, name, t);
+        else if (uid == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "%s: %s == 0, ignoring.", path, name);
+
+        *ret_uid = uid;
+        return 0;
+}
+
+static int read_login_defs(login_defs *ret_defs, const char *path) {
+        _cleanup_fclose_ FILE *f = NULL;
+        uid_t sys_uid_max = SYSTEM_UID_MAX;
+        gid_t sys_gid_max = SYSTEM_GID_MAX;
+        int r;
+
+        if (!path)
+                path = "/etc/login.defs";
+
+        r = fopen_unlocked(path, "re", &f);
+        if (r == -ENOENT)
+                goto assign;
+        if (r < 0)
+                return log_debug_errno(r, "Failed to open %s: %m", path);
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                char *t;
+
+                r = read_line(f, LINE_MAX, &line);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read %s: %m", path);
+                if (r == 0)
+                        break;
+
+                if ((t = first_word(line, "SYS_UID_MAX")))
+                        (void) parse_alloc_uid(path, "SYS_UID_MAX", t, &sys_uid_max);
+                else if ((t = first_word(line, "SYS_GID_MAX")))
+                        (void) parse_alloc_uid(path, "SYS_GID_MAX", t, &sys_gid_max);
+        }
+
+ assign:
+        *ret_defs = (login_defs) {
+                .system_uid_max = sys_uid_max,
+                .system_gid_max = sys_gid_max,
+        };
+        return 0;
+}
+#endif
+
+int acquire_login_defs(const login_defs **ret_defs) {
```

Comment:
since the function cannot fail, maybe just return the structure pointer directly, instead of using return-by-param?